consensus/propagation/have_wants.go

```diff
@@ -445,7 +445,7 @@ func (blockProp *Reactor) handleRecoveryPart(peer p2p.ID, part *proptypes.Recove
 	}
 	// the peer must always send the proposal before sending parts, if they did
 	// not this node must disconnect from them.
-	cb, parts, _, has := blockProp.getAllState(part.Height, part.Round, false)
+	cb, parts, _, has := blockProp.getAllState(part.Height, part.Round, part.Proof != nil)
```

Comment:
when receiving a recovery part, if the proof is nil, then it means it's not for catchup (we have the compact block) and we can verify the hash of the part to see if it's valid.